README.md
# WinNFSd

[![Build status](https://ci.appveyor.com/api/projects/status/github/winnfsd/winnfsd?svg=true)](https://ci.appveyor.com/project/MarcHarding/winnfsd-6y1xi/branch/master)

Introduction
------------
* Fork of WinNFSd_edited by ZeWarden(http://github.com/ZeWaren/WinNFSd_edited), based on WinNFSd by vincentgao (http://sourceforge.net/projects/winnfsd/).
* License: GPL.
* Runs on all major versions of Windows.

Description
--------------------
WinNFSd is a Network File System V3 (NFS) server for Windows.

You can use any NFS client to mount a directory of Windows and read/write files via NFS v3 protocol. It is useful when you usually access files of Windows on Linux and for especially for virtual machines, since it is much faster than shared folders.

You can also export any folder with an additional alias.

The export of multiple folders is also possible. Just put the shared foldes and an optional alias into a simple text file:

```
C:\path\to\a\mount > /alias
C:\path\to\another\mount /another-alias
```

Then start winnfsd.exe like this:
`WinNFSd.exe -pathFile C:\path\to\your\pathfile`


Usage
-------------------
```
=====================================================
WinNFSd 2.2.0
Network File System server for Windows
Copyright (C) 2005 Ming-Yang Kao
Edited in 2011 by ZeWaren
Edited in 2013 by Alexander Schneider (Jankowfsky AG)
Edited in 2014 2015 by Yann Schepens
Edited in 2016 by Peter Philipp (Cando Image GmbH), Marc Harding, Gregor Doltar 
=====================================================

Usage: WinNFSd.exe [-id <uid> <gid>] [-log on | off] [-pathFile <file>] [-addr <ip>] [-nfs3fhsize_32 on] [export path] [alias path]

At least a file or a path is needed
For example:
On Windows> WinNFSd.exe d:\work
On Linux> mount -t nfs 192.168.12.34:/d/work mount

For another example:
On Windows> WinNFSd.exe d:\work /exports
On Linux> mount -t nfs 192.168.12.34:/exports

Another example where WinNFSd is only bound to a specific interface:
On Windows> WinNFSd.exe -addr 192.168.12.34 d:\work /exports
On Linux> mount - t nfs 192.168.12.34: / exports

Use "." to export the current directory (works also for -filePath):
On Windows> WinNFSd.exe . /exports

Some NFS clients cannot handle 64 bit file handles (default). You may try with -nfs3fhsize_32 on. This will make them 32 bit wide.
```


CHANGELOG.md
## 2.2.1 (September 7,2016)

**BUG FIXES & IMPROVEMENTS:**
- Many memory leaks fixed (char * replaced with std::string) also memory was not freed
- Added nfs3fhsize_32 option. Some NFS clients do not like 64 bit handles
- When path is received it can be in form <path>/../<folder>. This introduced some problems. Path is now converted to canonical path,
- Not every reparse point is considered a link. Reparse point which mounts another volume is not a link,
- No longer case sensitive.

## 2.2.0 (August 6, 2016)

**BUG FIXES & IMPROVEMENTS:**

- Faster write speed (between 10 to 100 times faster) due to increased buffer size
- Added async write procedure for even faster writes for big files
- Fixed TCP implementation
- Added DUMP to PROG_MOUNT command for boot2docker and osx compatibility
- Added an alias option to the pathfile format (thanks to eidng8)
- Added an command line option to only listen on one interface (thanks to eidng8)

## 2.1.1 (July 10, 2016)

**BUG FIXES & IMPROVEMENTS:**

- Fix for slow READDIR
- Fixed odd log timestamp formatting

## 2.1.0 (June 23, 2016)

Many things have been fixed and WinNFSd works much better in everyday situations now.
This update is highly recommended for every user.

**BUG FIXES & IMPROVEMENTS:**

- Multiple, concurrent mounts (thanks to das-peter and Yasushi)
- SETATTR set size implemented. This fixes many, if not all problems with with git, node etc.
- SETATTR set mtime/atime implemented. This fixes touch calls [https://github.com/winnfsd/vagrant-winnfsd/issues/39]
- Symlinks now work across os [https://github.com/winnfsd/winnfsd/issues/10]
- Junctions and absolute symlinks are also supported (if their target resides inside the mount dir)
- Fix READDIR implementation [https://github.com/winnfsd/vagrant-winnfsd/issues/78]
- Files larger than 2GB are now supported (thanks to lepsiloon)
- miscellaneous fixes and memory leak fixes (thanks to Yasushi and philr)
- added a basic tests based on the Connectathon test suite

## Previous

The changelog starts with version 2.1.0. Any prior changes can be
seen by checking out the tagged releases and git commit messages.


src/FileTable.cpp
#include "FileTable.h"
#include "FileTree.h"
#include <string.h>
#include <io.h>
#include <stdio.h>
#include <windows.h>
#include <sys/stat.h>
#include "tree.hh"
#include "common.hh"
#include "Shlwapi.h"


static CFileTable g_FileTable;
static CFileTree g_FileTree;

CFileTable::CFileTable()
{
    m_pLastTable = m_pFirstTable = new FILE_TABLE;
    memset(m_pFirstTable, 0, sizeof(FILE_TABLE));
    m_nTableSize = 0;
    m_pCacheList = NULL;
}

CFileTable::~CFileTable()
{
    FILE_TABLE *pTable, *pTemp;
    unsigned int i;
    CACHE_LIST *pPrev;

    pTable = m_pFirstTable;

    while (pTable != NULL) { //free file table
        for (i = 0; i < TABLE_SIZE; i++) {
            if (!pTable->pItems[i]) {
                delete[] pTable->pItems[i];
            }
        }

        pTemp = pTable;
        pTable = pTable->pNext;
        delete pTemp;
    }

    while (m_pCacheList != NULL) { //free cache
        pPrev = m_pCacheList;
        m_pCacheList = m_pCacheList->pNext;
        delete pPrev;
    }
}
unsigned long CFileTable::GetIDByPath(char *path)
{
	unsigned char *handle;

	handle = GetHandleByPath(path);
	if (handle == NULL)
	{
		//printf("Can't find id for path %s\n", path);
		return 0;
	}
	return *(unsigned long *)handle;
}



unsigned char *CFileTable::GetHandleByPath(const char *path)
{
	tree_node_<FILE_ITEM> *node;

	node = g_FileTree.FindFileItemForPath(path);

	if (node == NULL) {
        AddItem(path);
		node = g_FileTree.FindFileItemForPath(path);
    }
	if (node == NULL) {
		return NULL;
	}

	return node->data.handle;
}

bool CFileTable::GetPathByHandle(unsigned char *handle, std::string& path)
{
	unsigned int id;
	tree_node_<FILE_ITEM>* node;

	id = *(unsigned int *)handle;
	// TODO : Not found item
	node = GetItemByID(id);
	if (node != NULL) {
		g_FileTree.GetNodeFullPath(node, path);
		return true;
	}
    return false;
}

tree_node_<FILE_ITEM>* CFileTable::FindItemByPath(char *path)
{
	tree_node_<FILE_ITEM>* node= NULL;/*
    CACHE_LIST *pCurr;

    unsigned int i, j, nPathLen;
    FILE_TABLE *pTable;

    nPathLen = strlen(path);
    pItem = NULL;

    pCurr = m_pCacheList;

    while (pCurr != NULL) { //search in cache
        if (nPathLen == pCurr->pItem->nPathLen) { //comparing path length is faster than comparing path
            if (strcmp(path, pCurr->pItem->path) == 0) { //compare path
                pItem = pCurr->pItem;  //path matched
                break;
            }
        }

        pCurr = pCurr->pNext;
    }

    if (pItem == NULL) { //not found in cache
        pTable = m_pFirstTable;

        for (i = 0; i < m_nTableSize; i += TABLE_SIZE) { //search in file table
            for (j = 0; j < TABLE_SIZE; j++) {
                if (i + j >= m_nTableSize) { //all items in file table are compared
                    return NULL;
                }

                if (nPathLen == pTable->pItems[j].nPathLen) { //comparing path length is faster than comparing path
                    if (strcmp(path, pTable->pItems[j].path) == 0) { //compare path
                        pItem = pTable->pItems + j;  //path matched
                        break;
                    }
                }
            }

            if (pItem != NULL) {
                break;
            }

            pTable = pTable->pNext;
        }
    }

    if (pItem != NULL) {
        //TODO IMPLEMENTED CACHE RIGHT
        //PutItemInCache(pItem);  //put the found item in cache
    }
	*/
	return node;
}

tree_node_<FILE_ITEM>* CFileTable::AddItem(const char *path)
{
    FILE_ITEM item;
	tree_node_<FILE_ITEM>* node;
    unsigned int nIndex;

	item.path = std::string(path);
	item.handle = new unsigned char[g_NFS3_FHSIZE];
	memset(item.handle, 0, g_NFS3_FHSIZE * sizeof(unsigned char));
	*(unsigned int *)item.handle = m_nTableSize;  //let its handle equal the index
	item.bCached = false;  //not in the cache

    if (m_nTableSize > 0 && (m_nTableSize & (TABLE_SIZE - 1)) == 0) {
        m_pLastTable->pNext = new FILE_TABLE;
        m_pLastTable = m_pLastTable->pNext;
        memset(m_pLastTable, 0, sizeof(FILE_TABLE));
    }

	//printf("\nAdd file %s for handle %i\n", path, *(unsigned int *)item.handle);

	g_FileTree.AddItem(path, item.handle);
	node = g_FileTree.FindFileItemForPath(path);
	if (node == NULL) {
		//printf("Can't find node just added %s\n", path);
	}

	m_pLastTable->pItems[nIndex = m_nTableSize & (TABLE_SIZE - 1)] = node;  //add the new item in the file table
    ++m_nTableSize;

	return node;  //return the pointer to the new item
}

tree_node_<FILE_ITEM>* CFileTable::GetItemByID(unsigned int nID)
{
	FILE_TABLE *pTable;
	unsigned int i;

	if (nID >= m_nTableSize) {
		return NULL;
	}

	pTable = m_pFirstTable;

	for (i = TABLE_SIZE; i <= nID; i += TABLE_SIZE) {
		pTable = pTable->pNext;
	}

	return pTable->pItems[nID + TABLE_SIZE - i];
}

void CFileTable::PutItemInCache(FILE_ITEM *pItem)
{
    CACHE_LIST *pPrev, *pCurr;
    int nCount;

    pPrev = NULL;
    pCurr = m_pCacheList;

    if (pItem->bCached) { //item is already in the cache
        while (pCurr != NULL) {
            if (pItem == pCurr->pItem) {
                if (pCurr == m_pCacheList) {  //at the first
                    return;
                }
                else {  //move to the first
                    pPrev->pNext = pCurr->pNext;
                    pCurr->pNext = m_pCacheList;
                    m_pCacheList = pCurr;
                    return;
                }
            }

            pPrev = pCurr;
            pCurr = pCurr->pNext;
        }
    }
    else {
        pItem->bCached = true;

        for (nCount = 0; nCount < 9 && pCurr != NULL; nCount++) { //seek to the end of the cache
            pPrev = pCurr;
            pCurr = pCurr->pNext;
        }

        if (nCount == 9 && pCurr != NULL) { //there are 10 items in the cache
            pPrev->pNext = NULL;  //remove the last
            pCurr->pItem->bCached = false;
        }
        else {
            pCurr = new CACHE_LIST;
        }

        pCurr->pItem = pItem;
        pCurr->pNext = m_pCacheList;
        m_pCacheList = pCurr;  //insert to the first
    }
}

bool CFileTable::RemoveItem(const char *path) {
   /* CACHE_LIST *pCurr;
    FILE_ITEM *pItem;
    unsigned int i, j, nPathLen;
    FILE_TABLE *pTable;
    int pItemIndex;

    nPathLen = strlen(path);
    pItem = NULL;

    bool foundDeletedItem = false;

    pTable = m_pFirstTable;

    for (i = 0; i < m_nTableSize; i += TABLE_SIZE) { //search in file table
        for (j = 0; j < TABLE_SIZE; j++) {
            if (i + j >= m_nTableSize) { //all items in file table are compared
                break;
            }

            if (!foundDeletedItem)
            {
                if (nPathLen == pTable->pItems[j].nPathLen) { //comparing path length is faster than comparing path
                    if (strcmp(path, pTable->pItems[j].path) == 0) { //compare path
                        foundDeletedItem = true;
                        memset(&(pTable->pItems[j]), 0, sizeof(FILE_ITEM));
                    }
                }
            }
        }

        pTable = pTable->pNext;
    }

    if (foundDeletedItem) {
		// we should not uncrement table size, because new file handle base on it
        //--m_nTableSize;
    }*/
	tree_node_<FILE_ITEM>* foundDeletedItem;
	foundDeletedItem = g_FileTree.FindFileItemForPath(path);
	if (foundDeletedItem != NULL) {
		// Remove from table
		FILE_TABLE *pTable;
		unsigned int i;
		unsigned int handle = *(unsigned int *)foundDeletedItem->data.handle;

		if (handle >= m_nTableSize) {
			//printf("File handle not found to remove : %s", path);
			return false;
		} else {

			pTable = m_pFirstTable;

			for (i = TABLE_SIZE; i <= handle; i += TABLE_SIZE) {
				pTable = pTable->pNext;
			}

			pTable->pItems[handle + TABLE_SIZE - i] = NULL;
		}
		// Remove from table end
		std::string item;
		g_FileTree.GetNodeFullPath(foundDeletedItem, item);
		g_FileTree.RemoveItem( item.c_str());
		return true;
	}
	else {
		//printf("File not found to remove : %s", path);
	}
    return false;
}

void CFileTable::RenameFile(char *pathFrom, char* pathTo)
{
	/*FILE_ITEM *pItem;

	pItem = g_FileTable.FindItemByPath(pathFrom);

	if (pItem == NULL) {
	} else {
		delete[] pItem->path;
		pItem->nPathLen = strlen(pathTo);
		pItem->path = new char[pItem->nPathLen + 1];
		strcpy_s(pItem->path, (pItem->nPathLen + 1), pathTo);  //replace the path by new one
	}
	*/
	g_FileTree.RenameItem(pathFrom, pathTo);
}

bool FileExists(char *path)
{
	char canPath[MAX_PATH];
	BOOL r;
	
	r = PathCanonicalize(canPath, path);
	if (!r)
		strncpy_s(canPath, path, MAX_PATH);

	r = PathFileExists(canPath);
	//I am not sure case compare really makes sense on windows
	return (r == false) ? FALSE : TRUE;
}

uint64_t GetFileID(char *path)
{
    return g_FileTable.GetIDByPath(path);
}

unsigned char *GetFileHandle(const char *path)
{
    return g_FileTable.GetHandleByPath(path);
}

void GetFilePath(unsigned char *handle, std::string &path)
{
    g_FileTable.GetPathByHandle(handle, path);
}

int RenameFile(char *pathFrom, char *pathTo)
{
	tree_node_<FILE_ITEM>* node;
    FILE_ITEM *pItem;

	node = g_FileTable.FindItemByPath(pathFrom);
    pItem = &(node->data);

    if (pItem == NULL) {
        return false;
    }

    errno_t errorNumber = rename(pathFrom, pathTo);

    if (errorNumber == 0) { //success
		g_FileTable.RenameFile(pathFrom, pathTo);
        return errorNumber;
    }
    else {
        return errorNumber;
    }
}

int RenameDirectory(char *pathFrom, char *pathTo)
{
	errno_t errorNumber = RenameFile(pathFrom, pathTo);

	const char* dotFile = "\\.";
	const char* backFile = "\\..";

	char* dotDirectoryPathFrom;
	char* dotDirectoryPathTo;
	char* backDirectoryPathFrom;
	char* backDirectoryPathTo;
	dotDirectoryPathFrom = (char *)malloc(strlen(pathFrom) + 1 + 3);
	strcpy_s(dotDirectoryPathFrom, (strlen(pathFrom) + 1), pathFrom);
	strcat_s(dotDirectoryPathFrom, (strlen(pathFrom) + 5), dotFile);

	dotDirectoryPathTo = (char *)malloc(strlen(pathTo) + 1 + 3);
	strcpy_s(dotDirectoryPathTo, (strlen(pathTo) + 1), pathTo);
	strcat_s(dotDirectoryPathTo, (strlen(pathTo) + 5), dotFile);

	backDirectoryPathFrom = (char *)malloc(strlen(pathFrom) + 1 + 4);
	strcpy_s(backDirectoryPathFrom, (strlen(pathFrom) + 1), pathFrom);
	strcat_s(backDirectoryPathFrom, (strlen(pathFrom) + 6), backFile);

	backDirectoryPathTo = (char *)malloc(strlen(pathTo) + 1 + 4);
	strcpy_s(backDirectoryPathTo, (strlen(pathTo) + 1), pathTo);
	strcat_s(backDirectoryPathTo, (strlen(pathTo) + 6), backFile);

	g_FileTable.RenameFile(dotDirectoryPathFrom, dotDirectoryPathTo);
	g_FileTable.RenameFile(backDirectoryPathFrom, backDirectoryPathTo);
	return errorNumber;


}

bool RemoveFile(char *path)
{
	int nMode = 0;
	nMode |= S_IREAD;
	nMode |= S_IWRITE;
	_chmod(path, nMode);

    if (remove(path) == 0){
        g_FileTable.RemoveItem(path);
        return true;
    }
    return false;
}

bool RemoveFolder(char *path)
{
	int nMode = 0;
	nMode |= S_IREAD;
	nMode |= S_IWRITE;
	_chmod(path, nMode);

    if (RemoveDirectory(path) != 0) {
        const char* dotFile = "\\.";
        const char* backFile = "\\..";

        char* dotDirectoryPath;
        char* backDirectoryPath;
        dotDirectoryPath = (char *)malloc(strlen(path) + 1 + 3);
        strcpy_s(dotDirectoryPath, (strlen(path) + 1), path);
        strcat_s(dotDirectoryPath, (strlen(path) + 5), dotFile);
        backDirectoryPath = (char *)malloc(strlen(path) + 1 + 4);
        strcpy_s(backDirectoryPath, (strlen(path) + 1), path);
        strcat_s(backDirectoryPath, (strlen(path) + 6), backFile);

        g_FileTable.RemoveItem(dotDirectoryPath);
        g_FileTable.RemoveItem(backDirectoryPath);
		g_FileTable.RemoveItem(path);
        return true;
    }
    return false;
}

void RemovePathFromFileTable(char *path)
{
    g_FileTable.RemoveItem(path);
}


src/winnfsd.cpp
#include "Socket.h"
#include "RPCServer.h"
#include "PortmapProg.h"
#include "NFSProg.h"
#include "MountProg.h"
#include "ServerSocket.h"
#include "DatagramSocket.h"
#include "common.hh"
#include <stdio.h>
#include <direct.h>
#include <iostream>
#include <fstream>
#include <vector>

#define SOCKET_NUM 3
enum
{
    PORTMAP_PORT = 111,
    MOUNT_PORT = 1058,
    NFS_PORT = 2049
};
enum
{
    PROG_PORTMAP = 100000,
    PROG_NFS = 100003,
    PROG_MOUNT = 100005
};

static unsigned int g_nUID, g_nGID;
static bool g_bLogOn;
static bool g_InputMode;
static char *g_sFileName;
static CRPCServer g_RPCServer;
static CPortmapProg g_PortmapProg;
static CNFSProg g_NFSProg;
static CMountProg g_MountProg;
size_t g_FHSIZE = 32;
size_t g_NFS3_FHSIZE = 64;

static void printUsage(char *pExe)
{
    printf("\n");
    printf("Usage: %s [-id <uid> <gid>] [-input] [-log on | off] [-nfs3fhsize_32 on | off] [-pathFile <file>] [-addr <ip>] [export path] [alias path]\n\n", pExe);
    printf("At least a file or a path is needed\n");
    printf("For example:\n");
    printf("On Windows> %s d:\\work\n", pExe);
    printf("On Linux> mount -t nfs 192.168.12.34:/d/work mount\n\n");
    printf("For another example:\n");
    printf("On Windows> %s d:\\work /exports\n", pExe);
    printf("On Linux> mount -t nfs 192.168.12.34:/exports\n\n");
    printf("Another example where WinNFSd is only bound to a specific interface:\n");
    printf("On Windows> %s -addr 192.168.12.34 d:\\work /exports\n", pExe);
    printf("On Linux> mount - t nfs 192.168.12.34: / exports\n\n");
    printf("Use \".\" to export the current directory (works also for -filePath):\n");
    printf("On Windows> %s . /exports\n", pExe);
}

static void printLine(void)
{
    printf("=====================================================\n");
}

static void printAbout(void)
{
    printLine();
    printf("WinNFSd {{VERSION}} [{{HASH}}]\n");
    printf("Network File System server for Windows\n");
    printf("Copyright (C) 2005 Ming-Yang Kao\n");
    printf("Edited in 2011 by ZeWaren\n");
    printf("Edited in 2013 by Alexander Schneider (Jankowfsky AG)\n");
	printf("Edited in 2014 2015 by Yann Schepens\n");
	printf("Edited in 2016 by Peter Philipp (Cando Image GmbH), Marc Harding, Gregor Doltar\n");
    printLine();
}

static void printHelp(void)
{
    printLine();
    printf("Commands:\n");
    printf("about: display messages about this program\n");
    printf("help: display help\n");
    printf("input: enable command input\n");
    printf("log on/off: display log messages or not\n");
    printf("list: list mounted clients\n");
    printf("refresh: refresh the mounted folders\n");
    printf("reset: reset the service\n");
    printf("quit: quit this program\n");
    printLine();
}

static void printCount(void)
{
    int nNum;

    nNum = g_MountProg.GetMountNumber();

    if (nNum == 0) {
        printf("There is no client mounted.\n");
    } else if (nNum == 1) {
        printf("There is 1 client mounted.\n");
    } else {
        printf("There are %d clients mounted.\n", nNum);
    }
}

static void printList(void)
{
    int i, nNum;

    printLine();
    nNum = g_MountProg.GetMountNumber();

    for (i = 0; i < nNum; i++) {
        printf("%s\n", g_MountProg.GetClientAddr(i));
    }

    printCount();
    printLine();
}

static void printConfirmQuit(void)
{
    printf("\n");
    printCount();
    printf("Are you sure to quit? (y/N): ");
}

static void mountPaths(std::vector<std::vector<std::string>> paths)
{
	int i;
	int numberOfElements = paths.size();

	for (i = 0; i < numberOfElements; i++) {
		char *pPath = (char*)paths[i][0].c_str();
		char *pPathAlias = (char*)paths[i][1].c_str();
		g_MountProg.Export(pPath, pPathAlias);  //export path for mount
	}
}

static void inputCommand(void)
{
	char command[20];

	printf("Type 'help' to see help\n\n");

	while (true) {
		fgets(command, 20, stdin);

		if (command[strlen(command) - 1] == '\n') {
			command[strlen(command) - 1] = '\0';
		}

		if (_stricmp(command, "about") == 0) {
			printAbout();
		} else if (_stricmp(command, "help") == 0) {
			printHelp();
		} else if (_stricmp(command, "log on") == 0) {
			g_RPCServer.SetLogOn(true);
		} else if (_stricmp(command, "log off") == 0) {
			g_RPCServer.SetLogOn(false);
		} else if (_stricmp(command, "list") == 0) {
			printList();
		} else if (_stricmp(command, "quit") == 0) {
			if (g_MountProg.GetMountNumber() == 0) {
				break;
			} else {
				printConfirmQuit();
				fgets(command, 20, stdin);

				if (command[0] == 'y' || command[0] == 'Y') {
					break;
				}
			}
		} else if (_stricmp(command, "refresh") == 0) {
			g_MountProg.Refresh();
		} else if (_stricmp(command, "reset") == 0) {
			g_RPCServer.Set(PROG_NFS, NULL);
		} else if (strcmp(command, "") != 0) {
			printf("Unknown command: '%s'\n", command);
			printf("Type 'help' to see help\n");
		}
	}
}

static void start(std::vector<std::vector<std::string>> paths)
{
	int i;
	CDatagramSocket DatagramSockets[SOCKET_NUM];
	CServerSocket ServerSockets[SOCKET_NUM];
	bool bSuccess;
	hostent *localHost;

	g_PortmapProg.Set(PROG_MOUNT, MOUNT_PORT);  //map port for mount
	g_PortmapProg.Set(PROG_NFS, NFS_PORT);  //map port for nfs
	g_NFSProg.SetUserID(g_nUID, g_nGID);  //set uid and gid of files

	mountPaths(paths);

	g_RPCServer.Set(PROG_PORTMAP, &g_PortmapProg);  //program for portmap
	g_RPCServer.Set(PROG_NFS, &g_NFSProg);  //program for nfs
	g_RPCServer.Set(PROG_MOUNT, &g_MountProg);  //program for mount
	g_RPCServer.SetLogOn(g_bLogOn);

	for (i = 0; i < SOCKET_NUM; i++) {
		DatagramSockets[i].SetListener(&g_RPCServer);
		ServerSockets[i].SetListener(&g_RPCServer);
	}

	bSuccess = false;

	if (ServerSockets[0].Open(PORTMAP_PORT, 3) && DatagramSockets[0].Open(PORTMAP_PORT)) { //start portmap daemon
		printf("Portmap daemon started\n");

		if (ServerSockets[1].Open(NFS_PORT, 10) && DatagramSockets[1].Open(NFS_PORT)) { //start nfs daemon
			printf("NFS daemon started\n");

			if (ServerSockets[2].Open(MOUNT_PORT, 3) && DatagramSockets[2].Open(MOUNT_PORT)) { //start mount daemon
				printf("Mount daemon started\n");
				bSuccess = true;  //all daemon started
			} else {
				printf("Mount daemon starts failed (check if port 1058 is not already in use ;) ).\n");
			}
		} else {
			printf("NFS daemon starts failed.\n");
		}
	} else {
		printf("Portmap daemon starts failed.\n");
	}


	if (bSuccess) {
		localHost = gethostbyname("");
		printf("Listening on %s\n", g_sInAddr);  //local address
		if (g_InputMode){
			inputCommand();  //wait for commands
		}
		else {
			//wait to be killed
			while (true)
				Sleep(100);
		}
	}

	for (i = 0; i < SOCKET_NUM; i++) {
		DatagramSockets[i].Close();
		ServerSockets[i].Close();
	}
}

int main(int argc, char *argv[])
{
    std::vector<std::vector<std::string>> pPaths;
    char *pPath = NULL;
	bool pathFile = false;

    WSADATA wsaData;

    printAbout();

    if (argc < 2) {
        pPath = strrchr(argv[0], '\\');
        pPath = pPath == NULL ? argv[0] : pPath + 1;
        printUsage(pPath);
        return 1;
    }

    g_nUID = g_nGID = 0;
    g_bLogOn = true;
    g_InputMode = false;
    g_sFileName = NULL;
	g_sInAddr = "0.0.0.0";

    for (int i = 1; i < argc; i++) {//parse parameters
        if (_stricmp(argv[i], "-id") == 0) {
            g_nUID = atoi(argv[++i]);
            g_nGID = atoi(argv[++i]);
        } else if (_stricmp(argv[i], "-input") == 0) {
            g_InputMode = true;
        } else if (_stricmp(argv[i], "-log") == 0) {
            g_bLogOn = _stricmp(argv[++i], "off") != 0;
        } else if (_stricmp(argv[i], "-nfs3fhsize_32") == 0) {
			if (!_stricmp(argv[++i], "on")) {
				g_NFS3_FHSIZE = g_FHSIZE;
				printf("NFS3_FHSIZE set to %d\n", g_NFS3_FHSIZE);
			}
        } else if (_stricmp(argv[i], "-addr") == 0) {
			g_sInAddr = argv[++i];
		} else if (_stricmp(argv[i], "-pathFile") == 0) {
            g_sFileName = argv[++i];

			if (g_MountProg.SetPathFile(g_sFileName) == false) {
                printf("Can't open file %s.\n", g_sFileName);
                return 1;
			} else {
				g_MountProg.Refresh();
				pathFile = true;
			}
        } else if (i == argc - 2) {
            pPath = argv[argc - 2];  //path is before the last parameter

            char *pCurPathAlias = argv[argc - 1]; //path alias is the last parameter

            if (pPath != NULL || pCurPathAlias != NULL) {
                std::vector<std::string> pCurPaths;
                pCurPaths.push_back(std::string(pPath));
                pCurPaths.push_back(std::string(pCurPathAlias));
                pPaths.push_back(pCurPaths);
            }

            break;
        } else if (i == argc - 1) {
            char *pPath = argv[argc - 1];  //path is the last parameter

            if (pPath != NULL) {
                char curPathAlias[MAXPATHLEN];
                strcpy_s(curPathAlias, pPath);
                char *pCurPathAlias = curPathAlias;

                std::vector<std::string> pCurPaths;
                pCurPaths.push_back(std::string(pPath));
                pCurPaths.push_back(std::string(pCurPathAlias));
                pPaths.push_back(pCurPaths);
            }

            break;
        }
    }

    HWND console = GetConsoleWindow();

    if (g_bLogOn == false && IsWindow(console)) {
        ShowWindow(console, SW_HIDE); // hides the window
    }

	if (pPaths.size() <= 0 && !pathFile) {
        printf("No paths to mount\n");
        return 1;
    }

    WSAStartup(0x0101, &wsaData);
    start(pPaths);
    WSACleanup();

    return 0;
}
